File: src/main/java/com/booknav/springbackend/services/AuthorService.java
package com.booknav.springbackend.services;

import com.booknav.springbackend.dto.AuthorDTO;
import com.booknav.springbackend.entities.Author;
import com.booknav.springbackend.repos.AuthorRepository;
import com.booknav.springbackend.utils.response.Response;
import org.modelmapper.ModelMapper;
import org.modelmapper.TypeMap;
import org.modelmapper.TypeToken;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

import java.lang.reflect.Type;
import java.util.List;
import java.util.Optional;

@Service
public class AuthorService {
    private final AuthorRepository authorRepository;
    private final ModelMapper modelMapper;

    @Autowired
    public AuthorService(AuthorRepository authorRepository, ModelMapper modelMapper) {
        this.authorRepository = authorRepository;
        this.modelMapper = modelMapper;
    }

    public ResponseEntity<Object> getAllAuthors() {
        try {
            List<Author> authors = authorRepository.findAll();
            Type listType = new TypeToken<List<AuthorDTO>>() {
            }.getType();

            TypeMap<Author, AuthorDTO> typeMap = modelMapper.getTypeMap(Author.class, AuthorDTO.class);
            if (typeMap == null) {
                typeMap = modelMapper.createTypeMap(Author.class, AuthorDTO.class);
            }
            typeMap.addMappings(mapping -> mapping.skip(AuthorDTO::setBooks));

            List<AuthorDTO> authorDTOS = modelMapper.map(authors, listType);

            return ResponseEntity.ok(Response.successRes("Authors retrieved successfully", authorDTOS));
        } catch (Exception e) {
            return ResponseEntity
                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body("Failed to retrieve authors");
        }
    }

    public ResponseEntity<Object> getAuthorById(int id) {
        try {
            Optional<Author> author = authorRepository.findById(id);
            if (author.isPresent()) {
                AuthorDTO authorDTO = modelMapper.map(author.get(), AuthorDTO.class);
                return ResponseEntity.ok(Response.successRes("Author retrieved successfully", authorDTO));
            } else {
                return ResponseEntity
                        .status(HttpStatus.NOT_FOUND)
                        .body(Response.errorRes("Author not found", null));
            }
        } catch (Exception e) {
            return ResponseEntity
                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Response.errorRes("Failed to retrieve the author", null));
        }
    }

    public ResponseEntity<Object> saveAuthor(AuthorDTO authorDTO, String message) {
        try {
            if (authorDTO.getAuthorId() == null) {
                Author author = modelMapper.map(authorDTO, Author.class);
                author = authorRepository.save(author);
                modelMapper.map(author, authorDTO);
                return ResponseEntity
                        .status(HttpStatus.CREATED)
                        .body(Response.successRes(message, authorDTO));
            } else {
                Optional<Author> author = authorRepository.findById(authorDTO.getAuthorId());
                if (author.isPresent()) {
                    modelMapper.map(authorDTO, author.get());
                    authorRepository.save(author.get());
                    modelMapper.map(author.get(), authorDTO);
                    return ResponseEntity.ok(Response.successRes(message, authorDTO));
                } else {
                    return ResponseEntity
                            .status(HttpStatus.NOT_FOUND)
                            .body(Response.errorRes("Author not found", null));
                }
            }
        } catch (Exception ex) {
            return ResponseEntity
                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Response.errorRes("Failed to save the author", null));
        }

    }

    public ResponseEntity<Object> deleteAuthorById(int id) {
        try {
            Optional<Author> author = authorRepository.findById(id);
            if (author.isPresent()) {
                authorRepository.deleteById(id);
                return ResponseEntity
                        .ok(Response.successRes("Author deleted successfully", "No data"));
            } else {
                return ResponseEntity
                        .status(HttpStatus.NOT_FOUND)
                        .body(Response.errorRes("Author not found", null));
            }
        } catch (Exception ex) {
            return ResponseEntity
                    .status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Response.errorRes("Failed to delete the author", null));
        }
    }
}

File: ASP Core - Backend/Services/Impl/AuthorService.cs
using aspcorebackend.Dtos;
using aspcorebackend.Models;
using aspcorebackend.Repositories;
using aspcorebackend.Utils.Response;
using AutoMapper;

namespace aspcorebackend.Services.Impl {
    public class AuthorService : IAuthorService {
        private readonly IAuthorRepository authorRepository;
        private readonly IMapper mapper;
        public AuthorService(IAuthorRepository authorRepository, IMapper mapper) {
            this.authorRepository = authorRepository;
            this.mapper = mapper;
        }

        public Response<object> GetAllAuthors(int? page) {
            try {
                List<Author> authors = authorRepository.GetAllAuthors(page);
                List<AuthorDTO> authorDTOs = authors.Select(a => {
                    a.Books = [];
                    return mapper.Map<AuthorDTO>(a);
                }).ToList();
                return ResponseMaker<object>.SuccessRes("Authors retrieved successfully", authorDTOs);
            } catch (Exception) {
                throw new Exception("Failed to retrieve authors");
            }
        }

        public Response<object> GetAuthorById(int id) {
            try {
                Author? author = authorRepository.GetAuthorWithBooksById(id);
                if (author == null) {
                    return ResponseMaker<object>.ErrorRes("Author not found", null);
                }
                AuthorDTO authorDTO = mapper.Map<AuthorDTO>(author);
                return ResponseMaker<object>.SuccessRes("Author retrieved successfully", authorDTO);
            } catch (Exception) {
                throw new Exception("Failed to retrieve the author");
            }
        }

        public Response<object> AddAuthor(AuthorDTO authorDTO) {
            try {
                Author author = mapper.Map<Author>(authorDTO);
                author = authorRepository.AddAuthor(author);
                authorDTO = mapper.Map<AuthorDTO>(author);
                return ResponseMaker<object>.SuccessRes("Author added successfully", authorDTO);
            } catch (Exception) {
                throw new Exception("Failed to save the author");
            }
        }

        public Response<object> UpdateAuthor(int id, AuthorDTO authorDTO) {
            try {
                Author? author = authorRepository.GetAuthorWithBooksById(id);
                if (author == null) {
                    return ResponseMaker<object>.ErrorRes("Author not found", null);
                }
                author.Firstname = authorDTO.Firstname;
                author.Lastname = authorDTO.Lastname;
                authorRepository.UpdateAuthor();
                authorDTO = mapper.Map<AuthorDTO>(author);
                return ResponseMaker<object>.SuccessRes("Author information updated successfully", authorDTO);
            } catch (Exception) {
                throw new Exception("Failed to update the author information");
            }
        }

        public Response<object> DeleteAuthor(int id) {
            try {
                Author? author = authorRepository.GetAuthorWithBooksById(id);
                if (author == null) {
                    return ResponseMaker<object>.ErrorRes("Author not found");
                }
                authorRepository.DeleteAuthor(author);
                return ResponseMaker<object>.SuccessRes("Author deleted successfully", null);
            } catch (Exception) {
                throw new Exception("Failed to delete the author");
            }
        }
    }
}
